File: agentic-crawler/src/agents/discoveryAgent.ts
import * as cheerio from 'cheerio';
import { type LLMClient, type CostTracker, DEFAULT_MODEL } from '../llm/client.js';

const FAQ_KEYWORDS = /faq|help|support|question|answer|guidance|query|enquir|knowledgebase/i;

const DISCOVERY_PROMPT = `You are a web navigation agent. Given a list of links from a webpage, identify all URLs that likely lead to FAQ or help content.

Return ONLY valid JSON in this format (no markdown, no explanation):
{"faqUrls": ["https://...", "https://..."]}

If no FAQ-related links are found, return: {"faqUrls": []}

FAQ indicators in URLs or link text: faq, help, support, questions, answers, guidance, query, enquiry, knowledgebase`;

/**
 * Discovers FAQ-related links on a page.
 * Results are restricted to the same hostname as pageUrl — never follows
 * cross-domain links, which prevents the crawler drifting to unrelated sites.
 *
 * Strategy:
 * 1. Fast heuristic pass — checks href + link text against FAQ keywords (free, no LLM)
 * 2. LLM pass — if heuristics find nothing, sends same-domain links to the LLM
 */
export async function discoverFaqLinks(
  pageUrl: string,
  rawHtml: string,
  llmClient: LLMClient,
  costTracker?: CostTracker
): Promise<string[]> {
  const pageHostname = new URL(pageUrl).hostname;
  const allLinks = extractLinks(pageUrl, rawHtml);

  // Restrict to same domain before anything else
  const sameDomainLinks = allLinks.filter((l) => {
    try { return new URL(l.href).hostname === pageHostname; } catch { return false; }
  });

  if (sameDomainLinks.length === 0) return [];

  // --- Pass 1: heuristic (no LLM cost) ---
  const heuristic = sameDomainLinks.filter(
    (l) => FAQ_KEYWORDS.test(l.href) || FAQ_KEYWORDS.test(l.text)
  );
  if (heuristic.length > 0) {
    return dedupe(heuristic.map((l) => l.href));
  }

  // --- Pass 2: LLM discovery (same-domain links only) ---
  const linkList = sameDomainLinks
    .slice(0, 100)
    .map((l) => `${l.text || '(no text)'} → ${l.href}`)
    .join('\n');

  const messages = [
    {
      role: 'user' as const,
      content: `${DISCOVERY_PROMPT}\n\nPage: ${pageUrl}\n\nLinks:\n${linkList}`,
    },
  ];

  try {
    const response = await llmClient.complete(messages, DEFAULT_MODEL);
    costTracker?.record(response.usage, DEFAULT_MODEL, `${pageUrl} (discovery)`);

    const cleaned = response.content.replace(/```(?:json)?\n?/g, '').trim();
    const parsed = JSON.parse(cleaned) as { faqUrls?: unknown[] };
    if (!Array.isArray(parsed.faqUrls)) return [];

    // Final guard: ensure LLM didn't hallucinate off-domain URLs
    return parsed.faqUrls
      .filter((u): u is string => typeof u === 'string')
      .filter((u) => { try { return new URL(u).hostname === pageHostname; } catch { return false; } });
  } catch (err) {
    console.error('[discovery] Failed to parse LLM response:', err);
    return [];
  }
}

function extractLinks(pageUrl: string, html: string): { href: string; text: string }[] {
  const $ = cheerio.load(html);
  const links: { href: string; text: string }[] = [];

  $('a[href]').each((_, el) => {
    const raw = $(el).attr('href') ?? '';
    const text = $(el).text().trim();

    if (!raw || raw.startsWith('#') || raw.startsWith('mailto:') || raw.startsWith('tel:')) {
      return;
    }
    try {
      const absolute = new URL(raw, pageUrl).href;
      links.push({ href: absolute, text });
    } catch {
      // skip malformed URLs
    }
  });

  return links;
}

function dedupe(urls: string[]): string[] {
  return [...new Set(urls)];
}

File: agentic-crawler-py/src/agents/discovery_agent.py
"""Discovery agent — finds FAQ/help links on a page (heuristic + LLM fallback)."""

from __future__ import annotations

import json
import re
from urllib.parse import urljoin, urlparse

from bs4 import BeautifulSoup

from ..llm.client import CostTracker, LLMClient, LLMMessage, DEFAULT_MODEL

FAQ_KEYWORDS = re.compile(
    r"faq|help|support|question|answer|guidance|query|enquir|knowledgebase", re.IGNORECASE
)

DISCOVERY_PROMPT = """You are a web navigation agent. Given a list of links from a webpage, identify all URLs that likely lead to FAQ or help content.

Return ONLY valid JSON in this format (no markdown, no explanation):
{"faqUrls": ["https://...", "https://..."]}

If no FAQ-related links are found, return: {"faqUrls": []}

FAQ indicators in URLs or link text: faq, help, support, questions, answers, guidance, query, enquiry, knowledgebase"""


async def discover_faq_links(
    page_url: str,
    raw_html: str,
    llm_client: LLMClient,
    cost_tracker: CostTracker | None = None,
) -> list[str]:
    """
    Discover FAQ-related links on a page.

    Strategy:
    1. Fast heuristic pass — checks href + link text against FAQ keywords (free, no LLM)
    2. LLM pass — if heuristics find nothing, sends same-domain links to the LLM
    """
    page_hostname = urlparse(page_url).hostname
    all_links = _extract_links(page_url, raw_html)

    same_domain = [
        lnk for lnk in all_links
        if _safe_hostname(lnk["href"]) == page_hostname
    ]

    if not same_domain:
        return []

    # Pass 1: heuristic (no LLM cost)
    heuristic = [
        lnk for lnk in same_domain
        if FAQ_KEYWORDS.search(lnk["href"]) or FAQ_KEYWORDS.search(lnk["text"])
    ]
    if heuristic:
        return list(dict.fromkeys(lnk["href"] for lnk in heuristic))

    # Pass 2: LLM discovery
    link_list = "\n".join(
        f"{lnk['text'] or '(no text)'} → {lnk['href']}"
        for lnk in same_domain[:100]
    )
    messages = [
        LLMMessage(
            role="user",
            content=f"{DISCOVERY_PROMPT}\n\nPage: {page_url}\n\nLinks:\n{link_list}",
        )
    ]

    try:
        response = await llm_client.complete(messages, DEFAULT_MODEL)
        if cost_tracker:
            cost_tracker.record(response.usage, DEFAULT_MODEL, f"{page_url} (discovery)")

        cleaned = re.sub(r"```(?:json)?\n?", "", response.content).strip()
        parsed = json.loads(cleaned)
        if not isinstance(parsed.get("faqUrls"), list):
            return []

        return [
            u for u in parsed["faqUrls"]
            if isinstance(u, str) and _safe_hostname(u) == page_hostname
        ]
    except Exception as e:
        print(f"[discovery] Failed to parse LLM response: {e}")
        return []


async def discover_general_links(
    page_url: str,
    raw_html: str,
) -> list[str]:
    """Return all same-domain links (used by general mode)."""
    page_hostname = urlparse(page_url).hostname
    all_links = _extract_links(page_url, raw_html)
    return list(dict.fromkeys(
        lnk["href"] for lnk in all_links
        if _safe_hostname(lnk["href"]) == page_hostname
    ))


FAQ_PATH_PATTERNS = ("/questions/", "/faq", "/help", "topic=")


def extract_faq_path_links(page_url: str, raw_html: str) -> list[str]:
    """Extract same-domain links whose URL matches FAQ path patterns (no LLM).

    Used as the primary link enqueue step in FAQ mode, replacing
    context.enqueue_links so we get visibility into what is found.
    """
    page_hostname = urlparse(page_url).hostname
    all_links = _extract_links(page_url, raw_html)
    results: list[str] = []
    for lnk in all_links:
        href = lnk["href"]
        if _safe_hostname(href) != page_hostname:
            continue
        lower = href.lower()
        if any(pat in lower for pat in FAQ_PATH_PATTERNS):
            results.append(href)
    return list(dict.fromkeys(results))


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _extract_links(page_url: str, html: str) -> list[dict[str, str]]:
    soup = BeautifulSoup(html, "lxml")
    links: list[dict[str, str]] = []
    for a in soup.find_all("a", href=True):
        raw = a["href"].strip()
        text = a.get_text(strip=True)
        if not raw or raw.startswith("#") or raw.startswith("mailto:") or raw.startswith("tel:"):
            continue
        try:
            absolute = urljoin(page_url, raw)
            links.append({"href": absolute, "text": text})
        except Exception:
            pass
    return links


def _safe_hostname(url: str) -> str | None:
    try:
        return urlparse(url).hostname
    except Exception:
        return None
